### ISSUE TITLE
Offload the 51B n-gram PLE table to SSD -- a lossless ~29 GB saving, already available as GGUF

### ISSUE THREAD

**Issue by evandhoffman** [evanwtf/local-llm#33]
Lead from [@EyalToledano](https://x.com/eyaltoledano/status/2092959261684846861),
2026-08-27. The headline is "fits on a 48gb MacBook Air", but the part that
matters here is a **lossless** memory win we can probably take without adopting
any of the rest.

## The two tricks are separable, and only one of them is free

He stacks two:

| trick | saving | cost |
|---|---|---|
| REAP pruning, 512 -> 288 experts | 97 GB -> 68 GB | **lossy**: HumanEval 93.9% -> 91.5% |
| **51B n-gram PLE table memmapped to SSD** | 97 GB -> ~68 GB | **lossless**: "logits identical to the ram-resident version in every way", bit-exact against the GPU path |

Together: 39 GB resident, 28 tok/s decode, 600 tok/s prefill, 6 s load.

**The second trick alone is worth ~29 GB and costs nothing.** That is the one to
chase. The pruning is a different proposition -- it removes 44% of the experts
(confirmed from our own GGUF: `qwen4exp.expert_count = 512`, `expert_used_count
= 10`) and produces a modified model.

## Why the PLE table is the right thing to offload

This matches an observation already made here: the 51B n-gram table is
**resident and does no arithmetic**. It is a lookup structure, ~100 bytes read
per token. Holding it in unified memory buys nothing but capacity pressure. It is
the single most offloadable component in the architecture, and our own metadata
names it explicitly:

```
qwen4exp.ple.ngram_size          3
qwen4exp.ple.heads_per_ngram     8
qwen4exp.ple.head_vocab_sizes    [arr x 16]
qwen4exp.ple.layers              [arr x 1]
```

## What ~29 GB would buy on this machine

Currently running `qwen38fnq3` = `UD-Q3_K_XL`, 83.8 GiB file, **88.5 GiB
resident**, 15/15 with Codex.

| if the PLE table were on disk | resident | consequence |
|---|---|---|
| current Q3 | ~59 GiB | ~53 GiB free for context, or room for a second model |
| **`UD-Q4_K_XL` (103.7 GiB)** | **~75 GiB** | **a strict upgrade: 4-bit instead of 3-bit, comfortably resident** |

Given that Q3 measured **28.4% faster than Q2** on agent wall time, moving up
another rung is the obvious next experiment, and PLE offload is what makes Q4
comfortable rather than marginal.

## The path that does not require his patch

His work is an **MLX** patch and is not obviously published. Our stack is
llama.cpp + shim + Codex.

But **AtomicChat already ships GGUFs that split the n-gram table so it can be
paged from SSD** -- the `-M64` suffix marks builds intended for 64 GB machines:

```
 79.1 GiB  Qwen3.8-Flash-Next-AD-3.84bpw-IQ4_XS-M64
 88.0 GiB  Qwen3.8-Flash-Next-AD-4.27bpw-Q4_K_M-M64
102.9 GiB  Qwen3.8-Flash-Next-AD-5.00bpw-Q5_K_M-M64
```

Those are GGUF, so they need no MLX and no custom patch. **Start there**: pull
`AD-4.27bpw-Q4_K_M-M64`, serve it on the existing PR #27742 build, and measure
*resident* memory rather than file size. File size is not the number -- the whole
point is that a large part of the file is meant to stay on disk.

## Do

1. Pull `AtomicChat/Qwen3.8-Flash-Next-GGUF` `AD-4.27bpw-Q4_K_M-M64` (88.0 GiB).
2. Serve it and record **resident vs file size**. If resident lands well below
   file size, the split is working.
3. Coherence check at temperature 0 before anything else (the #25 lesson).
4. Add as a separate backend and benchmark against `qwen38fnq3` with Codex.
5. If it works, the same question applies to `AD-5.00bpw-Q5_K_M-M64` (102.9 GiB),
   which would otherwise never fit.

## Two claims in the post NOT to carry over

**"28 tok/s decode" is not our metric.** Measured here today: Q3 is 28.4% faster
than Q2 on agent wall time *despite* being the larger, slower-per-token model,
because re-prefill dominates the agent loop (#14). Decode speed did not predict
that and would have predicted the opposite. His 600 tok/s prefill figure is the
more relevant one -- comparable to the ~483 t/s re-prefill measured here.

**"HumanEval 91.5% vs 93.9%" does not tell us the pruning is nearly free.**
HumanEval is single-function completion; every failure mode in this project is a
multi-turn agent loop failure. A 2.4pp HumanEval delta could mean nothing
agentically or a great deal -- that is exactly the gap #4 exists to record. Do
not treat REAP pruning as cheap on this evidence.

**Comment by evandhoffman** [comment]
Scope note: this issue covers the **specific** case — get the Qwen3.8-Flash-Next n-gram PLE table onto disk via AtomicChat's `-M64` GGUFs, and see whether `UD-Q4_K_XL` becomes comfortably resident.

The **general** question — what SSD offload costs in agent wall time across all three known implementations (ds4 expert streaming, the MLX PLE memmap, AtomicChat's split GGUFs), and where the break-even sits on a 128 GiB machine versus a 48 GB one — is now #34.

Do #34's step 1 first regardless: **nothing in this repo records the machine's actual NVMe read bandwidth or random-read latency.** Every claim about SSD offload being "fine" rests on numbers we have never measured, and the ~100 bytes-per-token access pattern is precisely the case where random-read latency matters more than sequential bandwidth.

**Comment by evandhoffman** [comment]
## The disk baseline exists now, and it does not say "free"

#34 step 1 measured this machine's NVMe for the first time (`benchmarks/disk/RESULTS.md`):

| pattern | median latency | throughput |
|---|---|---|
| sequential, 4 MiB | — | 9.45 GiB/s |
| random 1 MiB | 198 µs | 6.32 GiB/s |
| random 64 KiB | 90 µs | 1.00 GiB/s |
| random 4 KiB | **61 µs** | 0.10 GiB/s |

**Block size is what costs, not randomness.** Random 1 MiB reads reach 67% of sequential bandwidth; random 4 KiB reads reach 1.1%. The device is not penalising scattered access much — it is penalising *small* access.

That is bad news specifically for the trick this issue proposes.

### "~100 bytes per token" is not the cost

A 100-byte read cannot cost less than one block. The floor is **61 µs**, not 100 bytes' worth of bandwidth. So the question is not how many bytes the PLE table needs, it is **how many lookups a token needs** — and that number appears nowhere in the source claim or in this issue.

The arithmetic, using the source's own figures (28 tok/s decode, 600 tok/s prefill):

| lookups per token | decode cost | prefill cost |
|---|---|---|
| 1 | 0.2% | 3.7% |
| 2 | 0.3% | 7.3% |
| 8 | 1.4% | **29%** |
| 16 | 2.7% | **59%** |

**Decode is safe at any plausible lookup count.** Prefill is not.

### Why prefill is the wrong thing to slow down *here*

This is the part that makes the trade worse for this project than for the source's use case. #14 established that **re-prefill dominates agent wall time**, not decode — it is why `UD-Q3_K_XL` beat `UD-Q2_K_XL` by 28.4% while decoding *slower* per token, and why Claude Code on llama.cpp was 4.2x slower than Codex on the identical server. A single turn there re-prefilled 48,972 tokens in 101 s.

At 48,972 tokens and 8 lookups each, the added disk time is ~24 s on top of that turn. A technique whose cost lands almost entirely on prefill is close to the worst possible shape for an agent loop, whatever it does for chat.

### What this changes

**"Lossless" is a claim about output, not about time.** The source's "logits identical, verified bit-exact" is almost certainly true and is not in dispute. It says nothing about latency, and this issue has been carrying it as though it did — the framing "worth ~29 GB and costs nothing" is what needs revising.

The 29 GiB is real and worth having. The cost is unknown, bounded above by the table, and concentrated in exactly the phase this project cares most about.

**So the measurement plan needs one more step before the benchmark, and it is cheap:** get lookups-per-token. Either from the architecture (n-gram order should determine it) or empirically, by counting reads against the memmapped region during a short generation. If it comes back at 1–2, this is close to free and should be taken. At 8+, `UD-Q4_K_XL` resident is likely to beat `UD-Q3_K_XL`-plus-offload despite the better quant, and the comparison in step 2 should be built to show that rather than assume it.

Still worth doing — the AtomicChat `-M64` GGUF needs no MLX and no patch, and it is the cheapest real test available. Not started yet: it is a ~88 GiB download, and starting one during a timing batch is how an hour was lost on 2026-08-27.

**Comment by evandhoffman** [comment]
## Measured. The offload does not pay, and the memory saving was never available to win.

Both backends re-run on an identical stack — llama.cpp mainline `d7bd3bfca`, Codex 0.150.1, both through the shim. `qwen38fnq3` was re-run rather than compared against its earlier 895.8 s, because the engine, the client and the request path all moved today.

| task | q3 (3-bit, 83.8 GiB) | q4m64 (4-bit, 88.0 GiB) | delta |
|---|---|---|---|
| `mbox-strip-envelope` | 111.0 s | 103.1 s | −7% |
| `parser-mbox-quoting` | 155.3 s | 196.2 s | +26% |
| `storage-blob-put` | 140.5 s | 189.3 s | +35% |
| `mbox-scan` | 150.5 s | 195.2 s | +30% |
| `parser-date` | 437.8 s | 592.3 s | +35% |
| **suite** | **995.1 s** | **1,276.0 s** | **+28%** |

Correctness **15/15** and **16/16** — neither clears 90% at 95% confidence.

**+28% clears the ~26% resolution floor for a 3-trial suite, but only just.** What carries it is four of five tasks moving the same way at +26% to +35%; the fifth is −7%, inside the noise. Same standard as the Q3-vs-Q2 finding, same verdict shape: real, not comfortable.

### The mechanism was never going to help here

**`-M64` changes no tensors.** From the GGUF headers, against Unsloth's build:

```
split.tensors.count      1224   ==   1224
qwen4exp.ple.*           identical
split.count                 3   vs     33
```

Purely a re-sharding — so whether it saves memory is a question about llama.cpp's mmap, not about the model.

**And mmap already does it.** `vmmap` on the running server:

| config | RSS | physical footprint |
|---|---|---|
| default `-ngl 999` | 92.2 GiB | **4.8 GB** |
| PLE pinned to CPU | 91.9 GiB | **5.0 GB** |

RSS counts clean file-backed pages the kernel can drop at will; the footprint is memory the process actually owns. **All 88 GiB of weights are already evictable in both configs, with no flag and no special build.** I also tried `--override-tensor per_layer_token_embd\.weight=CPU` — it moved the tensor and changed nothing measurable, because on unified memory a CPU buffer and a Metal buffer are the same physical RAM.

So the ~29 GiB this issue set out to reclaim was not being held in the first place. There was no win available to take.

### The cost the source claim omits

The metadata answers what I flagged earlier as the deciding number: `ngram_size = 3`, **`heads_per_ngram = 8`**, 16 head offsets. Not the one or two lookups per token that would make this free.

Against the measured disk baseline (random 4 KiB = **61 µs**), a ~100-byte lookup cannot cost less than one block — ~9,600 random reads/s at 600 tok/s prefill against 25,559 IOPS. Feasible, not free, and landing on prefill, which #14 established dominates agent wall time here.

**"Lossless" is a claim about output, not about time.** Almost certainly true of the logits; silent about the clock.

### Recommendation

**Keep `UD-Q3_K_XL`.** The better quant costs 28% of the wall clock and buys no measured correctness.

Worth keeping the 4-bit build as evidence that a 125B model at 4-bit loads and runs clean on a 128 GiB machine — that is a useful fact about the ceiling even though it is not the backend to run.

Closing this. **#34 step 3 stays open and is now better motivated:** block size is what costs on this device — random 1 MiB reads hit 6.32 GiB/s against 0.10 GiB/s for 4 KiB — and MoE expert blocks are hundreds of KiB to a few MiB. ds4's `--ssd-streaming-*` reads at the size the hardware is good at. That is a different proposition from an n-gram table read 8 times per token, and this result says nothing against it.
